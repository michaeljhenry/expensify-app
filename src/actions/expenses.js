import uuid from 'uuid';

export const addExpense = ({description='', note='',amount = 0, createdAt = 0} = {} ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

// REMOVE EXPENSE ACTION

export const removeExpense = ({id} = {} ) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT EXPENSE ACTION

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});