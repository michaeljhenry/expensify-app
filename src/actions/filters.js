

export const setTextFilter = (text='') => ({
    type: 'SET_TEXT_FILTER',
    text
});

// SORT BY AMOUNT ACTION

export const sortByAmount = () => ({
    type: 'SORT_BY_AMOUNT',
});

// SORT BY DATE ACTION

export const sortByDate = () => ({
    type: 'SORT_BY_DATE'
});

// SET START DATE

export const setStartDate = (startDate) => ({
    type: 'SET_START_DATE',
    startDate
});

// SET END DATE

export const setEndDate = (endDate) => ({
    type: 'SET_END_DATE',
    endDate
});