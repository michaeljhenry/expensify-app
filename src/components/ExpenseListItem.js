import React from 'react';
import moment from 'moment';
import {Link} from 'react-router-dom';
import numeral from 'numeral';




const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
        <Link className = 'list-item' to = {`edit/${id}`}>
        <div>
            <h3 className = 'list-item__title'>{description}</h3>  
            <span className = 'list-item__subtitle'>{moment(createdAt).format('MMMM Do, YYYY')}</span>
        </div>
        <h3 className = 'list-item__data'>{numeral(amount/100).format('$0,0.00')}</h3>

        </Link>
)

export default ExpenseListItem;

// const ExpenseListItem = ({id, description, amount, createdAt}) => (
//     <div>
//         <h3>{description}</h3>
//         <p>{amount} - {createdAt}</p>
//         <button>Remove</button>
//     </div>
// )


// export default ExpenseListItem;