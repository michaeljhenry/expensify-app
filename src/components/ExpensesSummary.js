import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import visibleExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import moment from 'moment'


export class ExpensesSummary extends React.Component {
    render() {
        return(
            <div>
                <p>{expensesTotal(this.props.expenses)}</p>
                <p>{visibleExpenses(this.props.expenses, this.props.filters)}</p>
            </div>
        );
    };
};

const mapStateToProps = (state,props) => ({
    expenses: state.expenses,
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    visibleExpenses: (expenses, filters) => dispatch(visibleExpenses(expenses, filters)),
    expensesTotal: (expenses) => dispatch(expensesTotal(expenses))
});
export default connect(mapStateToProps,mapDispatchToProps)(ExpensesSummary);