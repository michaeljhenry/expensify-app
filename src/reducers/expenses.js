
// Expenses Reducer
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id); // deconstruct the expenses object that the filter function iterates over?
/*             return state.filter(() => {
                entry => entry.id !== action.id
            });  THIS IS HOW I DID IT..IT WORKED*/
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }

                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;

        default:
            return state;
    }
};

