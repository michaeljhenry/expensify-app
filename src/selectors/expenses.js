import moment from 'moment'

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expenses) => {
        const createdAtMoment = moment(expenses.createdAt);
        // Explaining this code in case I look at it and get confused later...
        // They provide us a list of expenses and we want to see if we should include each item.
        // They enter a startDate that they want to filter from. Every expense before that date will be filtered out and not shown.
        // Similar process followed for endDate. textMatch makes sure the text is inside the string.
        // If all these are true, it won't be filtered out and will then be added to the sort function.
        const startDateMatch = startDate? startDate.isSameOrBefore(createdAtMoment, 'day'): true; 
        const endDateMatch = endDate? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expenses.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if((sortBy === 'date') || (sortBy === '')) {
            return a.createdAt < b.createdAt? 1 : -1; // most recent expenses come first
        }
        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
