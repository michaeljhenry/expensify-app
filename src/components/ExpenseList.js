import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

export const ExpenseList = (props) => ( // exporting this unconnected version because we want to use it for testing
    <div>
    {
        props.expenses.length === 0? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => {
                return <ExpenseListItem key = {expense.id} {...expense}/> // Why didn't I just pass in {expense} instead of {...expense} ? Aren't they both an object with the same info ?
            })
        )
    }

        
    </div>
);

const mapStateToProps = (state) => { // maps the store state to the component props
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

