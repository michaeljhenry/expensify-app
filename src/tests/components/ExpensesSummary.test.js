import {ExpensesSummary} from '../../components/ExpensesSummary';
import {shallow} from 'enzyme';
import React from 'react';
import expenses from '../fixtures/expenses';
import {filters, altFilters} from '../fixtures/filters';
import moment from 'moment';

let visibleExpenses, expensesTotal, wrapper;

beforeEach(() => {
    visibleExpenses = jest.fn();
    expensesTotal = jest.fn();

    wrapper = shallow(
        <ExpensesSummary
            visibleExpenses={visibleExpenses}
            expensesTotal = {expensesTotal}
            expenses = {expenses}
            filters = {filters}
        />)
});


test('should display the correct amount of expenses and correct toal', () => {
    expect(wrapper).toMatchSnapshot();
}); 