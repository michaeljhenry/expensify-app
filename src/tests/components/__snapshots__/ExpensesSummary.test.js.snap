// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the correct amount of expenses and correct toal 1`] = `
<div>
  <p>
    114195
  </p>
</div>
`;
